storm-server/src/main/java/org/apache/storm/daemon/nimbus/Nimbus.java

```diff
@@ -4723,4 +4754,192 @@ public IScheduler getForcedScheduler() {
+        private final Histogram supervisorsUsedMem = registerHistogram.apply("supervisors:used-mem");
+        private final Histogram supervisorsUsedCpu = registerHistogram.apply("supervisors:used-CPU");
+        private final Histogram supervisorsFragmentedMem = registerHistogram.apply("supervisors:fragmented-mem");
+        private final Histogram supervisorsFragmentedCpu = registerHistogram.apply("supervisors:fragmented-CPU");
+
+        //Topology metrics distribution
+        private final Histogram topologiesNumTasks = registerHistogram.apply("topologies:num-tasks");
+        private final Histogram topologiesNumExecutors = registerHistogram.apply("topologies:num-executors");
+        private final Histogram topologiesNumWorker = registerHistogram.apply("topologies:num-workers");
+        private final Histogram topologiesUptime = registerHistogram.apply("topologies:uptime-secs");
+        private final Histogram topologiesReplicationCount = registerHistogram.apply("topologies:replication-count");
+        private final Histogram topologiesRequestedMemOnHeap = registerHistogram.apply("topologies:requested-mem-on-heap");
+        private final Histogram topologiesRequestedMemOffHeap = registerHistogram.apply("topologies:requested-mem-off-heap");
+        private final Histogram topologiesRequestedCpu = registerHistogram.apply("topologies:requested-CPU");
+        private final Histogram topologiesAssignedMemOnHeap = registerHistogram.apply("topologies:assigned-mem-on-heap");
+        private final Histogram topologiesAssignedMemOffHeap = registerHistogram.apply("topologies:assigned-mem-off-heap");
+        private final Histogram topologiesAssignedCpu = registerHistogram.apply("topologies:assigned-CPU");
+
+        ClusterSummaryMetricSet() {
+            //Break the code if out of sync to thrift protocol
+            assert ClusterSummary._Fields.values().length == 3
+                && ClusterSummary._Fields.findByName("supervisors") == ClusterSummary._Fields.SUPERVISORS
+                && ClusterSummary._Fields.findByName("topologies") == ClusterSummary._Fields.TOPOLOGIES
+                && ClusterSummary._Fields.findByName("nimbuses") == ClusterSummary._Fields.NIMBUSES;
+
+            final CachedGauge<ClusterSummary> cachedSummary = new CachedGauge<ClusterSummary>(CACHING_WINDOW, TimeUnit.SECONDS) {
+                @Override
+                protected ClusterSummary loadValue() {
+                    try {
+                        if (active) {
+                            ClusterSummary newSummary = getClusterInfoImpl();
+                            LOG.info("the new summary is {}", newSummary);
+                            //This is ugly but I can't think of a better way to update histogram only once per caching
+                            // It also kind of depends on the implementation that gauges gets updated before histograms
+                            updateHistogram(newSummary);
```

Comment:
If I understand this correctly, updateHistogram will only be called every `CACHING_WINDOW` i.e. 5s. And histogram is `new SlidingTimeWindowReservoir(CACHING_WINDOW / 2, TimeUnit.SECONDS)`, which means it only collects data in most recent 2.5s. The histogram will be dealing with basically the same data in the most of time.

Reply:
Yes, because I want to collect the distribution of value (let's say the assigned cpu) across all entities (daemons, or topologies) only within one clusterSummary (so it's the distribution of assigned cpu among all supervisors at one moment, snapshotted in ClusterSummary). It doesn't make sense to me to accumulate both across entities and time.